CIP-ledger-evolution/README.md

```diff
@@ -0,0 +1,302 @@
+
+#### Changes that *do not* require a CIP
+
+* performance improvements
+* bug fixes
+* ledger events
+* changes to the protocol parameter values (this is a governance issue, see [CIP-1694])
+
+### Types of change
+
+- [Ledger features](#ledger-features)
+  - [Ledger-script interface](#ledger-script-interface)
+- [Ledger intra-era switch](#ledger-intra-era-switch)
+- [Ledger protocol](#ledger-protocol)
+- [Ledger eras](#ledger-eras-1)
+- [Serialization](#serialization)
+
+#### Ledger features
+
+This is a very broad category.
+Any conceptual group of changes to the ledger semantics,
+involving a single idea or a group of closely related ideas
+(such as adding, modifying, or removing existing features)
+is a `ledger-feature`.
+This category excludes bug fixes and changes to the [ledger protocol](#ledger-protocol).
+Ledger features are assumed to be included in a future ledger era.
+
+To become active, the ledger feature must be included in an active [ledger era](#ledger-eras-1).
+
+##### Ledger-script interface
+
+Any conceptual group of changes to the script context is a `ledger-script-interface`.
+This is a shared type with Plutus, so the process in [CIP-35] MUST also be followed.
+
+To become active, the interface change must be included in an active [ledger era](#ledger-eras-1).
+
+#### Ledger intra-era switch
+
+Sometimes a ledger change is so small and targeted that it can be implemented by
+switching on the major protocol version (and not introducing a new ledger era).
+These must be used *very* sparingly, as they can quickly lead to unintended complexity and
+unclear semantics[^2].
+A ledger intra-era switch does not necessarily require a CIP;
+sometimes it might be more appropriate to just include it in a [ledger era](#ledger-eras-1) CIP.
+
+When a hard fork involves only intra-era switching, a new ledger era can be avoided and we
+refer to the hard fork as an intra-era hard fork.
+
+To become active, the ledger intra-era switch must be included in an active [ledger era](#ledger-eras-1).
+
+#### Ledger protocol
+
+The `ledger-protocol` category corresponds closely to block header validation.
+It corresponds precisely to all the functionality in the Shelley ledger specification
+not captured by the `BBODY` and `TICK` transitions.
+This mostly involves handling the nonces for Ouroboros, the VRF checks,
+the KES certificates, and block leadership checks.
+
+Historically, the ledger protocols have been specified alongside the ledger eras.
+The `TPraos` ledger protocol was specified alongside Shelley, and
+`Praos` was specified alongside Babbage.
+
+To become active, a ledger protocol change MUST be
+specified with the STS framework and MUST be included in the software update for a soft/hard fork.
+If the fork coincides with a ledger era, the specifications can be combined.
+
+#### Ledger eras
+
+A `ledger-era` is a collection of ledger-features, bug fixes, intra-era switches, and other minor changes
```

Comment:
Yep, you are totally right, the "ledger era" is very much like a release. The reasons I wanted to include them into the CIP process are:

* many non-trivial details come up when one goes to combine features
* our specifications are done in units of ledger eras
* someone needs to decide when and how to bundle the features, so this is something else to open up to everyone
* some hard fork changes are so silly and small that they do not need their own CIP, but could be tacked on to an overall bundling of features

---

Your "unrestricted burning" example is definitely a `ledger-feature`. It may or may not be worth giving the full STS change details without also knowing the other changes that it will be bundled with. but certainly some prose and clarity around what it changes in the STS is very helpful. cryptographic details and wire formats are probably best left for the `ledger-era` CIP that includes it.

Reply:
Aha, I think I see a point where we differ. I think you're envisaging that, regardless of the existence of CIPs for individual features, you would submit a final CIP when you want to do a new era that bundles up everything.

This is a bit different to how I did it in CIP-35: the idea there was that the composition of "releases" was not in scope for the CIP process. So there might be CIPs A B and C that we want to do which all require a new Plutus Version, and D E and F which don't, and we might decide to bundle A C and D together into PlutusV3. And we don't have to make a synthesis CIP that announces that. We might need to write a _specification_ for it, but that's different.

I think you _could_ do something similar for the ledger, and just announce that the composition of eras is out of scope. Then you'd never have a CIP for an era, you'd just have CIPs for features, and the ledge team would decide how to release them in eras (or not).

Reply:
Starting from these assumptions:

* the ledger needs a CIP to legitimize itself as a CIP category
* ledger specs (which encompass eras) [should](https://github.com/input-output-hk/cardano-ledger/issues/3178) be in the CIP process
* the "types of changes" in CIP-35 are helpful
* I could not imagine strict delineations like CIP-35 for ledger

I came up with:

* Provide a document with all the concepts that we use in the ledger team when discussing changes.
* Separate conceptual units of change (which, realistically, can be almost anything) from the "bundling" and "spec producing" work.
* mention the events and the ledger protocol, as those are actually pretty contained
* write specific guidelines for serialization and spec writing

Maybe there is a way to turn this on its side and do something totally different? I'm prepared to have a total eclipse of the heart. :)